### PR TITLE
CC-14610 Unzer Payment System Provider

### PR DESCRIPTION
**Pull request by kkicheglovspryker**
- Developer(s): @kkicheglovspryker, @michbeck

- Ticket: https://spryker.atlassian.net/browse/CC-14610

- Release Group: https://release.spryker.com/release-groups/view/4066

- PR Overview: https://release.spryker.com/core/CC-14610

- merge: squash

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   UnzerRestApi              | minor (new)                 |  Unzer                     |

-----------------------------------------

##### Change log

Initial Release

- Introduced Glue API endpoints for payment functionality of Unzer payment service provider.

### CODE REVIEW

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/UnzerNotificationMapper.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+     * @param \Generated\Shared\Transfer\RestUnzerNotificationAttributesTransfer $restUnzerNotificationAttributesTransfer
+     * @param \Generated\Shared\Transfer\UnzerNotificationTransfer $unzerNotificationTransfer
+     *
+     * @return \Generated\Shared\Transfer\UnzerNotificationTransfer
+     */
+    public function mapRestUnzerNotificationAttributesTransferToUnzerNotificationTransfer(
+        RestUnzerNotificationAttributesTransfer $restUnzerNotificationAttributesTransfer,
+        UnzerNotificationTransfer $unzerNotificationTransfer
+    ): UnzerNotificationTransfer {
+        return $unzerNotificationTransfer->fromArray($restUnzerNotificationAttributesTransfer->toArray(), true);
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/UnzerNotificationMapper.php]: fromArray / toArray is not a real mnapping
this class could be deleted
code could be moved directly to a place where it's used

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    /**
+     * @var \SprykerEco\Glue\UnzerRestApi\Dependency\UnzerRestApiToUnzerClientInterface
+     */
+    protected $unzerClient;
+
+    /**
+     * @var \Spryker\Glue\GlueApplication\Rest\JsonApi\RestResourceBuilderInterface
+     */
+    protected $resourceBuilder;
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php]:
```suggestion
    protected $restResourceBuilder;
```

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        return $this->createRestSuccessResponse($unzerNotificationTransfer->getIsProcessedOrFail());
+    }
+
+    /**
+     * @param bool $isProcessed
+     *
+     * @return \Spryker\Glue\GlueApplication\Rest\JsonApi\RestResponseInterface
+     */
+    protected function createRestSuccessResponse(bool $isProcessed): RestResponseInterface
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php]: it look like it not always `sucess`

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi;
+
+use Spryker\Glue\Kernel\AbstractBundleConfig;
+
+class UnzerRestApiConfig extends AbstractBundleConfig
+{
+    /**
+     * @var string
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php:20]: `@api` tag is missing

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * @param \Spryker\Glue\Kernel\Container $container
+     *
+     * @return \Spryker\Glue\Kernel\Container
+     */
+    public function provideDependencies(Container $container): Container
+    {
+        parent::provideDependencies($container);
+
+        $this->addUnzerClient($container);
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]:
```suggestion
        $container = $this->addUnzerClient($container);
```

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    public const CLIENT_UNZER = 'CLIENT_UNZER';
+
+    /**
+     * @param \Spryker\Glue\Kernel\Container $container
+     *
+     * @return \Spryker\Glue\Kernel\Container
+     */
+    public function provideDependencies(Container $container): Container
+    {
+        parent::provideDependencies($container);
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]:
```suggestion
        $container = parent::provideDependencies($container);
```

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        return $container;
+    }
+
+    /**
+     * @param \Spryker\Glue\Kernel\Container $container
+     *
+     * @return void
+     */
+    protected function addUnzerClient(Container $container): void
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]:
```suggestion
    protected function addUnzerClient(Container $container): Container
```

return container at this method

File: tooling.yml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+architecture-sniffer:
+  priority: 2
+
+code-sniffer:
+    level: 1
```
Comment (DmitryLymarenko) [inline tooling.yml]:
```suggestion
    level: 2
```

**Review by DmitryLymarenko — CHANGES_REQUESTED** [review]

**Comment by DmitryLymarenko** [comment]
Release type should be `major (new)`
`merge: merge` -> `squash`

**Comment by DmitryLymarenko** [comment]
Looks like we have a constraint to Unzer because of client method call

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php (hunk 4 above)
Comment (kkicheglovspryker) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php:20]: fixed

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php (hunk 5 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]: fixed.

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php (hunk 6 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]: fixed.

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php (hunk 7 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiDependencyProvider.php]: fixed.

File: tooling.yml (hunk 8 above)
Comment (lausen) [reply tooling.yml]: fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/UnzerNotificationResourceRoutePlugin.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi\Plugin;
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/UnzerNotificationResourceRoutePlugin.php]:
```suggestion
namespace SprykerEco\Glue\UnzerRestApi\Plugin\GlueApplication;
```

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+use Spryker\Glue\Kernel\AbstractBundleConfig;
+
+class UnzerRestApiConfig extends AbstractBundleConfig
+{
+    /**
+     * @api
+     *
+     * @var string
+     */
+    public const UNZER_NOTIFICATION = 'unzer-notification';
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: resources are usually named in the plural form
`wishlists`
`configurable-bundle-templates`
`unzer-notifications`
etc

File: src/SprykerEco/Glue/UnzerRestApi/Controller/UnzerNotificationController.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+use Generated\Shared\Transfer\RestUnzerNotificationAttributesTransfer;
+use Spryker\Glue\GlueApplication\Rest\JsonApi\RestResponseInterface;
+use Spryker\Glue\GlueApplication\Rest\Request\Data\RestRequestInterface;
+use Spryker\Glue\Kernel\Controller\AbstractController;
+
+/**
+ * @method \SprykerEco\Glue\UnzerRestApi\UnzerRestApiFactory getFactory()
+ */
+class UnzerNotificationController extends AbstractController
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Controller/UnzerNotificationController.php:17]:
```suggestion
class UnzerNotificationsController extends AbstractController
```
following the same logic
but class naming at Processor folder should be left in singular form

File: src/SprykerEco/Shared/UnzerRestApi/Transfer/unzer_rest_api.transfer.xml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<transfers xmlns="spryker:transfer-01" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="spryker:transfer-01 http://static.spryker.com/transfer-01.xsd">
+    <transfer name="RestUnzerNotificationAttributes">
+    </transfer>
+
+    <transfer name="UnzerNotification">
+        <property name="isProcessed" type="bool"/>
+    </transfer>
+
+    <transfer name="RestUnzerNotificationAttributes">
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Shared/UnzerRestApi/Transfer/unzer_rest_api.transfer.xml]:
```suggestion
    <transfer name="RestUnzerNotificationsAttributes">
```

**Review by DmitryLymarenko — CHANGES_REQUESTED** [review]

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php (hunk 10 above)
Comment (DmitryLymarenko) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: also constant name should be `RESOURCE_UNZER_NOTIFICATIONS`

take a look at some rest api module
may be `ConfigurableBundlesRestApi`

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/UnzerNotificationResourceRoutePlugin.php (hunk 9 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/Plugin/UnzerNotificationResourceRoutePlugin.php]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php (hunk 10 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Controller/UnzerNotificationController.php (hunk 11 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/Controller/UnzerNotificationController.php:17]: Fixed.

File: src/SprykerEco/Shared/UnzerRestApi/Transfer/unzer_rest_api.transfer.xml (hunk 12 above)
Comment (lausen) [reply src/SprykerEco/Shared/UnzerRestApi/Transfer/unzer_rest_api.transfer.xml]: Fixed. Also removed empty first Entity with same name attribue.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace SprykerEco\Glue\UnzerRestApi\Plugin\GlueApplication;
+
+use Generated\Shared\Transfer\RestUnzerNotificationsAttributesTransfer;
+use Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceRouteCollectionInterface;
+use Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceRoutePluginInterface;
+use Spryker\Glue\Kernel\AbstractPlugin;
+use SprykerEco\Glue\UnzerRestApi\UnzerRestApiConfig;
+
+class UnzerNotificationResourceRoutePlugin extends AbstractPlugin implements ResourceRoutePluginInterface
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php]:
```suggestion
class UnzerNotificationsResourceRoutePlugin extends AbstractPlugin implements ResourceRoutePluginInterface
```

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+{
+    /**
+     * @var string
+     */
+    protected const CONTROLLER_NAME = 'unzer-notification';
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php]:
```suggestion
    protected const CONTROLLER_NAME = 'unzer-notifications';
```

this code is currently not workable
please check it before passing to review

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+use Spryker\Glue\Kernel\AbstractBundleConfig;
+
+class UnzerRestApiConfig extends AbstractBundleConfig
+{
+    /**
+     * @api
+     *
+     * @var string
+     */
+    public const UNZER_NOTIFICATIONS = 'unzer-notification';
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: not fixed

`RESOURCE_UNZER_NOTIFICATIONS = 'unzer-notifications';`

**Review by DmitryLymarenko — CHANGES_REQUESTED** [review]

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php (hunk 15 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php (hunk 14 above)
Comment (lausen) [reply src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueApplication/UnzerNotificationResourceRoutePlugin.php]: Changes done.

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+use Spryker\Glue\Kernel\AbstractBundleConfig;
+
+class UnzerRestApiConfig extends AbstractBundleConfig
+{
+    /**
+     * @api
+     *
+     * @var string
+     */
+    public const UNZER_NOTIFICATIONS = 'unzer-notifications';
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php]: still not fixed

**Review by DmitryLymarenko — APPROVED** [review]

File: phpstan.neon [hunk 17]
```diff
@@ -0,0 +1,9 @@
+parameters:
+    level: 8
+    paths:
+        - src/
+    bootstrapFiles:
+        - tests/bootstrap.php
+    checkGenericClassInNonGenericObjectType: false
```
Comment (asmarovydlo) [inline phpstan.neon]: Please fix according to https://spryker.slack.com/archives/C7N7YEHK8/p1650969879005379

File: src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi;
+
+use Spryker\Glue\Kernel\AbstractBundleConfig;
+
+class UnzerRestApiConfig extends AbstractBundleConfig
+{
+    /**
+     * @api
```
Comment (asmarovydlo) [inline src/SprykerEco/Glue/UnzerRestApi/UnzerRestApiConfig.php:18]: Please add specificaiton.

File: psalm-report.json [hunk 19]
```diff
@@ -0,0 +1,210 @@
+{
+    "error": [],
+    "warning": [],
+    "deprecation": [
+        {
+            "severity": "info",
+            "line_from": 43,
+            "line_to": 43,
+            "type": "DeprecatedClass",
+            "message": "Class Spryker\\Glue\\GlueApplication\\Rest\\Request\\Data\\RestRequestInterface is marked as deprecated",
```
Comment (asmarovydlo) [inline psalm-report.json]: This module implements the deprecated API concept, was that planned or agreed with someone?

File: composer.json [hunk 20]
```diff
@@ -0,0 +1,75 @@
+{
+    "name": "spryker-eco/unzer-rest-api",
+    "type": "library",
+    "description": "UnzerRestApi module",
+    "license": "MIT",
+    "require": {
+        "php": ">=7.4",
+        "spryker-eco/unzer": "dev-feature/cc-14610/dev-unzer-psp as 1.0",
+        "spryker/glue-application": "^1.0.0",
+        "spryker/glue-json-api-convention-extension": "^0.1.0",
```
Comment (asmarovydlo) [inline composer.json:9]: Stable modules should not release of unstable, please update release type to `minor(new)`

File: LICENSE [hunk 21]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (asmarovydlo) [inline LICENSE]:
```suggestion
```

File: psalm-report.json (hunk 19 above)
Comment (asmarovydlo) [reply psalm-report.json]: Please re-generate the report, deprecations should go.

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php [hunk 22]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi\Processor\Notification;
+
+use Generated\Shared\Transfer\GlueResponseTransfer;
```
Comment (asmarovydlo) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Notification/UnzerNotificationProcessor.php:10]: Missing in transfer definition.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueJsonApiConventionExtension/UnzerNotificationResource.php [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi\Plugin\GlueJsonApiConventionExtension;
+
+use Generated\Shared\Transfer\GlueResourceMethodCollectionTransfer;
+use Generated\Shared\Transfer\GlueResourceMethodConfigurationTransfer;
```
Comment (asmarovydlo) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/GlueJsonApiConventionExtension/UnzerNotificationResource.php:11]: Missing in transfer definition.

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
Please update the module description in the changelogs, now it says nothing about API. Use smth similar to what you have now in the readme.

**Review by asmarovydlo — APPROVED** [review]

**Comment by kkicheglovspryker** [comment]
image

**Review by asmarovydlo — APPROVED** [review]

**Review by asmarovydlo — APPROVED** [review]

**Pull request merged**